
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Comprehensive project documentation for scVIDR.">
      
      
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Interrogation of VAE Using Ridge Regression in Portal Hepatocyte Response Prediction - scVIDR</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.342714a4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../assets/css/custom.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#interrogation-of-vae-using-ridge-regression-in-portal-hepatocyte-response-prediction" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="scVIDR" class="md-header__button md-logo" aria-label="scVIDR" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            scVIDR
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Interrogation of VAE Using Ridge Regression in Portal Hepatocyte Response Prediction
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../installation/getting-started/" class="md-tabs__link">
        
  
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../API/API-introduction/" class="md-tabs__link">
        
  
  
    
  
  API

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../tutorials-introduction/" class="md-tabs__link">
        
  
  
    
  
  Tutorial

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="scVIDR" class="md-nav__button md-logo" aria-label="scVIDR" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    scVIDR
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/getting-started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../API/API-introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials-introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tutorial
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#importing-required-libraries-and-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Importing Required Libraries and Functions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-loading-and-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      Data Loading and Preprocessing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#training-and-loading-the-vae-model-for-portal-hepatocyte-dose-response-prediction" class="md-nav__link">
    <span class="md-ellipsis">
      Training and Loading the VAE Model for Portal Hepatocyte Dose-Response Prediction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ridge-regression-analysis-on-latent-representations" class="md-nav__link">
    <span class="md-ellipsis">
      Ridge Regression Analysis on Latent Representations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#figure-4b-visualization-of-top-gene-scores" class="md-nav__link">
    <span class="md-ellipsis">
      Figure 4B: Visualization of Top Gene Scores
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#figure-4c-gene-enrichment-analysis-and-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      Figure 4C: Gene Enrichment Analysis and Visualization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#figure-4d-gene-enrichment-analysis-and-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      Figure 4D: Gene Enrichment Analysis and Visualization
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#figure-4e-logistic-fit-of-median-pathway-scores" class="md-nav__link">
    <span class="md-ellipsis">
      Figure 4E: Logistic Fit of Median Pathway Scores
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="interrogation-of-vae-using-ridge-regression-in-portal-hepatocyte-response-prediction">Interrogation of VAE Using Ridge Regression in Portal Hepatocyte Response Prediction<a class="headerlink" href="#interrogation-of-vae-using-ridge-regression-in-portal-hepatocyte-response-prediction" title="Permanent link">&para;</a></h1>
<p>This analysis utilizes a trained scVIDR model to investigate portal hepatocyte responses to TCDD by calculating gene scores through ridge regression on latent dimensions. Genes with high scores indicate significant transcriptomic changes under increasing doses. Enrichr analysis highlights pathways involved in TCDD-induced hepatotoxicity, including cytochrome P450 and fatty acid oxidation. A sigmoid function models the relationship between pathway scores and doses, showing a strong dose-response relationship.</p>
<h2 id="importing-required-libraries-and-functions">Importing Required Libraries and Functions<a class="headerlink" href="#importing-required-libraries-and-functions" title="Permanent link">&para;</a></h2>
<p>Imports the necessary libraries and modules for processing single-cell gene expression data, model implementation, statistical analysis, and interpretation. It includes tools for machine learning, visualization, statistical tests, and pathway enrichment analysis to facilitate the analysis of scVIDR models and ridge regression.</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;../vidr/&#39;</span><span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="c1">#Import VIDR functions</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">vidr</span><span class="w"> </span><span class="kn">import</span> <span class="n">VIDR</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">PCAEval</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCAEval</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">utils</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>

<span class="c1">#Import important modules</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">linalg</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">spatial</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="kn">import</span> <span class="n">AnnData</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">sparse</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">statannotations.Annotator</span><span class="w"> </span><span class="kn">import</span> <span class="n">Annotator</span>

<span class="c1">#Interpetation of model</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.linear_model</span><span class="w"> </span><span class="kn">import</span> <span class="n">ElasticNet</span><span class="p">,</span> <span class="n">Lars</span><span class="p">,</span> <span class="n">Ridge</span><span class="p">,</span> <span class="n">LinearRegression</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">mean_absolute_error</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">gseapy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">gp</span>
</code></pre></div>
<h2 id="data-loading-and-preprocessing">Data Loading and Preprocessing<a class="headerlink" href="#data-loading-and-preprocessing" title="Permanent link">&para;</a></h2>
<p>Loads single-cell gene expression data, filters out immune cells to retain specific liver cell types of interest, and preprocesses the dataset. Preprocessing steps include normalizing gene expression, applying a logarithmic transformation, and identifying the top 5,000 highly variable genes for further analysis. Finally, it converts the dose values into categorical strings to facilitate downstream analyses.</p>
<div class="highlight"><pre><span></span><code><span class="n">doses</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">30</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;../data/nault2021_multiDose.h5ad&quot;</span><span class="p">)</span>

<span class="c1">#Remove Immune Cells</span>
<span class="n">cell_types_of_int</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Hepatocytes - central&quot;</span><span class="p">,</span> <span class="s2">&quot;Hepatocytes - portal&quot;</span><span class="p">,</span> <span class="s2">&quot;Cholangiocytes&quot;</span><span class="p">,</span> <span class="s2">&quot;Stellate Cells&quot;</span><span class="p">,</span> <span class="s2">&quot;Portal Fibroblasts&quot;</span><span class="p">,</span> <span class="s2">&quot;Endothelial Cells&quot;</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cell_types_of_int</span><span class="p">)]</span>

<span class="c1">#Preporces data</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span>

<span class="c1">#Make dose categorical</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dose&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Dose&quot;</span><span class="p">]]</span>
</code></pre></div>
<h2 id="training-and-loading-the-vae-model-for-portal-hepatocyte-dose-response-prediction">Training and Loading the VAE Model for Portal Hepatocyte Dose-Response Prediction<a class="headerlink" href="#training-and-loading-the-vae-model-for-portal-hepatocyte-dose-response-prediction" title="Permanent link">&para;</a></h2>
<p>Prepares data specific to portal hepatocytes, a cell type of interest, by splitting it into training and test sets. It initializes and optionally trains a Variational Autoencoder (VAE) model to predict gene expression changes in response to TCDD doses. The trained model is then loaded for further analysis.</p>
<div class="highlight"><pre><span></span><code><span class="n">model_name</span> <span class="o">=</span> <span class="s2">&quot;VAE&quot;</span>
<span class="n">celltypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">])</span>
<span class="n">cell</span> <span class="o">=</span> <span class="s1">&#39;Hepatocytes - portal&#39;</span>
<span class="n">train_adata</span><span class="p">,</span> <span class="n">test_adata</span> <span class="o">=</span> <span class="n">prepare_cont_data</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="s2">&quot;dose&quot;</span><span class="p">,</span> <span class="s2">&quot;Dose&quot;</span><span class="p">,</span> <span class="n">cell</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"> Predicting </span><span class="si">{</span><span class="n">cell</span><span class="si">}</span><span class="s2"> Dioxin Dose Response</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">VIDR</span><span class="p">(</span><span class="n">train_adata</span><span class="p">,</span> <span class="n">linear_decoder</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="c1"># model.train(</span>
<span class="c1"># max_epochs=100,</span>
<span class="c1"># batch_size=128,</span>
<span class="c1"># early_stopping=True,</span>
<span class="c1"># early_stopping_patience=25)</span>
<span class="c1"># model.save(f&quot;../../data/VAE_Cont_Prediction_Dioxin_5000g_{cell}.pt/&quot;, overwrite = True)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;../data/VAE_Cont_Prediction_Dioxin_5000g_</span><span class="si">{</span><span class="n">cell</span><span class="si">}</span><span class="s2">.pt/&quot;</span><span class="p">,</span> <span class="n">train_adata</span><span class="p">)</span>
</code></pre></div>
<h2 id="ridge-regression-analysis-on-latent-representations">Ridge Regression Analysis on Latent Representations<a class="headerlink" href="#ridge-regression-analysis-on-latent-representations" title="Permanent link">&para;</a></h2>
<p>This series of steps leverages the trained model to predict the dose-response, generate synthetic samples, and apply ridge regression.</p>
<p>Generates dose-response predictions for the specified cell type and doses, with delta capturing the response vector.
<div class="highlight"><pre><span></span><code><span class="n">regression</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">dr_dict</span><span class="p">,</span> <span class="n">delta</span><span class="p">,</span> <span class="n">reg</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
    <span class="n">ctrl_key</span><span class="o">=</span><span class="s2">&quot;0.0&quot;</span><span class="p">,</span>
    <span class="n">treat_key</span><span class="o">=</span><span class="s2">&quot;30.0&quot;</span><span class="p">,</span>
    <span class="n">cell_type_to_predict</span><span class="o">=</span><span class="n">cell</span><span class="p">,</span>
    <span class="n">regression</span> <span class="o">=</span> <span class="n">regression</span><span class="p">,</span>
    <span class="n">continuous</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">doses</span> <span class="o">=</span> <span class="n">doses</span><span class="p">)</span>
</code></pre></div></p>
<p>Extracts latent space representations, identifies the range for each dimension, and creates random samples within this range.
<div class="highlight"><pre><span></span><code><span class="n">latent</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
<span class="n">mins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">latent</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">maxes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">latent</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">rand_samp</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="p">(</span><span class="n">mn</span><span class="p">,</span><span class="n">mx</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mins</span><span class="p">,</span> <span class="n">maxes</span><span class="p">):</span>
    <span class="n">rand_samp</span> <span class="o">+=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">mn</span><span class="p">,</span><span class="n">mx</span><span class="p">,</span>  <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="mi">100000</span><span class="p">))]</span>
<span class="n">rand_samp</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rand_samp</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
</code></pre></div></p>
<p>Simulates gene expression data using the model's generative module from the random samples in latent space.
<div class="highlight"><pre><span></span><code><span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda:0&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>
<span class="n">gen_samp</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">module</span><span class="o">.</span><span class="n">generative</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">rand_samp</span><span class="p">)</span><span class="o">.</span><span class="n">float</span><span class="p">())[</span><span class="s2">&quot;px&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
</code></pre></div></p>
<p>Calculates weights for each sample based on cosine similarity to the delta vector, emphasizing samples aligned with the dose-response.
<div class="highlight"><pre><span></span><code><span class="n">weights</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">rand_samp</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="n">weights</span> <span class="o">+=</span> <span class="p">[</span><span class="mi">1</span><span class="o">/</span><span class="n">spatial</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">cosine</span><span class="p">(</span><span class="n">rand_samp</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="n">delta</span><span class="p">)]</span>
</code></pre></div></p>
<p>Fits a ridge regression model to predict the generative outputs using the latent space samples and their weights.
<div class="highlight"><pre><span></span><code><span class="n">reg</span> <span class="o">=</span> <span class="n">Ridge</span><span class="p">()</span>
<span class="n">reg</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">rand_samp</span><span class="p">,</span> <span class="n">gen_samp</span><span class="p">,</span> <span class="n">sample_weight</span><span class="o">=</span><span class="n">weights</span><span class="p">)</span>
</code></pre></div></p>
<p>Generates new random samples within the latent space range for testing the ridge regression model.
<div class="highlight"><pre><span></span><code><span class="n">test_samp</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="p">(</span><span class="n">mn</span><span class="p">,</span><span class="n">mx</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mins</span><span class="p">,</span> <span class="n">maxes</span><span class="p">):</span>
    <span class="n">test_samp</span> <span class="o">+=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">mn</span><span class="p">,</span><span class="n">mx</span><span class="p">,</span>  <span class="n">size</span> <span class="o">=</span> <span class="p">(</span><span class="mi">20000</span><span class="p">))]</span>
<span class="n">test_samp</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">test_samp</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
</code></pre></div></p>
<p>Uses the generative module to simulate gene expression data for the test samples.
<div class="highlight"><pre><span></span><code><span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda:0&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>
<span class="n">test_gen_samp</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">module</span><span class="o">.</span><span class="n">generative</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">test_samp</span><span class="p">)</span><span class="o">.</span><span class="n">float</span><span class="p">())[</span><span class="s2">&quot;px&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
</code></pre></div></p>
<p>Evaluates the ridge regression model's performance on the test samples by calculating the R<sup>2</sup> score.
<div class="highlight"><pre><span></span><code><span class="n">reg</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">test_samp</span><span class="p">,</span><span class="n">test_gen_samp</span><span class="p">)</span>
</code></pre></div></p>
<div class="codehilite"><pre><span></span><code><span class="mf">0.6035409897482676</span>
</code></pre></div>

<p>Calculates normalized gene weights by dividing the regression coefficients by their L2 norms to scale the impact of each gene.
<div class="highlight"><pre><span></span><code><span class="n">gene_weights</span> <span class="o">=</span> <span class="n">reg</span><span class="o">.</span><span class="n">coef_</span>
<span class="n">gene_norms</span> <span class="o">=</span> <span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">reg</span><span class="o">.</span><span class="n">coef_</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">gene_weights_norm</span> <span class="o">=</span> <span class="n">gene_weights</span> <span class="o">/</span> <span class="n">gene_norms</span>
</code></pre></div>
Computes the importance of each feature (gene) by projecting the normalized gene weights onto the delta vector.</p>
<div class="highlight"><pre><span></span><code><span class="n">feature_importance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">gene_weights_norm</span><span class="p">,</span> <span class="n">delta</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]))</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
</code></pre></div>
<p>Identifies the top 100 genes contributing to the dose-response, sorted by their calculated feature importance.
<div class="highlight"><pre><span></span><code><span class="n">train_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">feature_importance</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]][:</span><span class="mi">100</span><span class="p">]</span>
</code></pre></div></p>
<div class="codehilite"><pre><span></span><code><span class="nx">Index</span><span class="p">([</span><span class="err">&#39;</span><span class="nx">Egfr</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Ttc39c</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Ghr</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cps1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Kynu</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Slco1b2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cdh18</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Fmo3</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Cyp3a25</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cyp7b1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Nrg1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Mug2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Aass</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Ldb2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Dlc1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cyp1a1</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Ankrd44</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Airn</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Il1r1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Reln</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="mi">9030622</span><span class="nx">O22Rik</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Slc7a2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Nfe2l2</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Esrrg</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Meis2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Pvt1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="mi">4930578</span><span class="nx">G10Rik</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Gab2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cyp2c29</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Pde4d</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Cyp2e1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Gm20663</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Mecom</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Gm2163</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Fli1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Aox3</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Heg1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cfh</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Tiparp</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cyp1a2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Gfra1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Sdk1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Pck1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Slc16a10</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Auts2</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Camk2d</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Cyp3a11</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Robo1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Ehd3</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Plpp1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Prkg1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Rnf213</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Col4a1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Acsl1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Col4a2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Nrp1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Uox</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Slc8a1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Apbb2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Tmcc3</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Prickle1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Car3</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Tcf4</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Gm26917</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Rnf125</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Pitpnc1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Samd4</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">B3galt1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Slc38a4</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Palld</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Bmp6</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Mrc1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Hao1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Maf</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Jam2</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Fam129b</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Tshz2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Jak1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Syne1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Abcc4</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Prex2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Tead1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Hmcn1</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Abi1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">C9</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Adamts9</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Fkbp5</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">F8</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Eng</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Tmem51</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Plxna4</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Ces3b</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Dapk2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Fndc3b</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Rftn1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Kcnb1</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Ahrr</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Maml2</span><span class="err">&#39;</span><span class="p">,</span><span class="w"> </span><span class="err">&#39;</span><span class="nx">Malat1</span><span class="err">&#39;</span><span class="p">,</span>
<span class="w">       </span><span class="err">&#39;</span><span class="nx">Selenop</span><span class="err">&#39;</span><span class="p">],</span>
<span class="w">      </span><span class="nx">dtype</span><span class="p">=</span><span class="err">&#39;</span><span class="nx">object</span><span class="err">&#39;</span><span class="p">)</span>
</code></pre></div>

<h2 id="figure-4b-visualization-of-top-gene-scores">Figure 4B: Visualization of Top Gene Scores<a class="headerlink" href="#figure-4b-visualization-of-top-gene-scores" title="Permanent link">&para;</a></h2>
<p>Creates a bar plot showing the top 20 genes with the highest scVIDR gene scores, emphasizing their contribution to the dose-response prediction. The genes are displayed in descending order of importance.</p>
<div class="highlight"><pre><span></span><code><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">train_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">feature_importance</span><span class="p">)][</span><span class="o">-</span><span class="mi">20</span><span class="p">:][::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">feature_importance</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">feature_importance</span><span class="p">)][</span><span class="o">-</span><span class="mi">20</span><span class="p">:][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span> <span class="o">=</span> <span class="mi">90</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">12</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;scVIDR Gene Score&quot;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">18</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;../figures/4B.svg&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span> <span class="o">=</span> <span class="s2">&quot;tight&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div>
<p><img alt="png" src="../Figure4_files/Figure4_14_0.png" /></p>
<h2 id="figure-4c-gene-enrichment-analysis-and-visualization">Figure 4C: Gene Enrichment Analysis and Visualization<a class="headerlink" href="#figure-4c-gene-enrichment-analysis-and-visualization" title="Permanent link">&para;</a></h2>
<p>Identifies the top 100 genes based on their scores, performs an Enrichr pathway enrichment analysis using the WikiPathways 2019 Mouse database, and visualizes the top 15 enriched pathways with significant adjusted p-values. The bar plot highlights pathways critical to TCDD response.</p>
<div class="highlight"><pre><span></span><code><span class="n">glist</span> <span class="o">=</span> <span class="n">train_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">feature_importance</span><span class="p">)][</span><span class="o">-</span><span class="mi">100</span><span class="p">:][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">glist</span> <span class="o">=</span> <span class="n">glist</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">enr</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">glist</span><span class="p">,</span>
                 <span class="n">gene_sets</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;WikiPathways_2019_Mouse&quot;</span><span class="p">],</span>
                 <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="c1"># don&#39;t forget to set organism to the one you desired! e.g. Yeast</span>
                 <span class="n">description</span><span class="o">=</span><span class="s1">&#39;test_name&#39;</span><span class="p">,</span>
                 <span class="n">outdir</span><span class="o">=</span><span class="s1">&#39;test/enrichr_kegg&#39;</span><span class="p">,</span>
                 <span class="c1"># no_plot=True,</span>
                 <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.5</span> <span class="c1"># test dataset, use lower value from range(0,1)</span>
                <span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="n">enr</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</code></pre></div>
<div style="overflow-x: auto; width: 100%; max-width: 100vw;">
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene_set</th>
      <th>Term</th>
      <th>Overlap</th>
      <th>P-value</th>
      <th>Adjusted P-value</th>
      <th>Old P-value</th>
      <th>Old Adjusted P-value</th>
      <th>Odds Ratio</th>
      <th>Combined Score</th>
      <th>Genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Tryptophan metabolism WP79</td>
      <td>5/44</td>
      <td>0.000003</td>
      <td>0.000140</td>
      <td>0</td>
      <td>0</td>
      <td>26.802969</td>
      <td>344.396446</td>
      <td>KYNU;CYP1A2;CYP1A1;CYP2E1;CYP7B1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Fatty Acid Omega Oxidation WP33</td>
      <td>3/7</td>
      <td>0.000004</td>
      <td>0.000140</td>
      <td>0</td>
      <td>0</td>
      <td>153.835052</td>
      <td>1905.147874</td>
      <td>CYP1A2;CYP1A1;CYP2E1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Oxidation by Cytochrome P450 WP1274</td>
      <td>4/40</td>
      <td>0.000047</td>
      <td>0.001046</td>
      <td>0</td>
      <td>0</td>
      <td>22.990741</td>
      <td>229.196526</td>
      <td>CYP1A2;CYP1A1;CYP2E1;CYP7B1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Metapathway biotransformation WP1251</td>
      <td>5/141</td>
      <td>0.000714</td>
      <td>0.011957</td>
      <td>0</td>
      <td>0</td>
      <td>7.648607</td>
      <td>55.412606</td>
      <td>CYP1A2;CYP1A1;FMO3;CYP2E1;CYP7B1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Ethanol metabolism resulting in production of ...</td>
      <td>2/10</td>
      <td>0.001085</td>
      <td>0.014540</td>
      <td>0</td>
      <td>0</td>
      <td>50.744898</td>
      <td>346.391772</td>
      <td>CYP2E1;NFE2L2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Estrogen metabolism WP1264</td>
      <td>2/14</td>
      <td>0.002166</td>
      <td>0.019906</td>
      <td>0</td>
      <td>0</td>
      <td>33.823129</td>
      <td>207.504274</td>
      <td>CYP1A2;CYP1A1</td>
    </tr>
    <tr>
      <th>6</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Keap1-Nrf2 WP1245</td>
      <td>2/14</td>
      <td>0.002166</td>
      <td>0.019906</td>
      <td>0</td>
      <td>0</td>
      <td>33.823129</td>
      <td>207.504274</td>
      <td>MAF;NFE2L2</td>
    </tr>
    <tr>
      <th>7</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Focal Adhesion WP85</td>
      <td>5/185</td>
      <td>0.002377</td>
      <td>0.019906</td>
      <td>0</td>
      <td>0</td>
      <td>5.766082</td>
      <td>34.838505</td>
      <td>RELN;COL4A2;COL4A1;SELENOP;EGFR</td>
    </tr>
    <tr>
      <th>8</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Complement and Coagulation Cascades WP449</td>
      <td>3/62</td>
      <td>0.003704</td>
      <td>0.027578</td>
      <td>0</td>
      <td>0</td>
      <td>10.400664</td>
      <td>58.225148</td>
      <td>F8;CFH;C9</td>
    </tr>
    <tr>
      <th>9</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Focal Adhesion-PI3K-Akt-mTOR-signaling pathway...</td>
      <td>6/324</td>
      <td>0.005757</td>
      <td>0.038573</td>
      <td>0</td>
      <td>0</td>
      <td>3.930550</td>
      <td>20.271055</td>
      <td>GHR;RELN;COL4A2;COL4A1;EGFR;JAK1</td>
    </tr>
    <tr>
      <th>10</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>MicroRNAs in Cardiomyocyte Hypertrophy WP1560</td>
      <td>3/82</td>
      <td>0.008073</td>
      <td>0.049173</td>
      <td>0</td>
      <td>0</td>
      <td>7.759755</td>
      <td>37.395832</td>
      <td>CAMK2D;NRG1;PRKG1</td>
    </tr>
    <tr>
      <th>11</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>Nuclear receptors in lipid metabolism and toxi...</td>
      <td>2/30</td>
      <td>0.009829</td>
      <td>0.054876</td>
      <td>0</td>
      <td>0</td>
      <td>14.483965</td>
      <td>66.951564</td>
      <td>CYP1A2;CYP2E1</td>
    </tr>
    <tr>
      <th>12</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>EGFR1 Signaling Pathway WP572</td>
      <td>4/178</td>
      <td>0.012269</td>
      <td>0.063234</td>
      <td>0</td>
      <td>0</td>
      <td>4.723659</td>
      <td>20.787206</td>
      <td>ABI1;GAB2;EGFR;JAK1</td>
    </tr>
    <tr>
      <th>13</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>ESC Pluripotency Pathways WP339</td>
      <td>3/118</td>
      <td>0.021399</td>
      <td>0.099196</td>
      <td>0</td>
      <td>0</td>
      <td>5.320932</td>
      <td>20.455914</td>
      <td>SELENOP;EGFR;JAK1</td>
    </tr>
    <tr>
      <th>14</th>
      <td>WikiPathways_2019_Mouse</td>
      <td>ErbB signaling pathway WP1261</td>
      <td>2/46</td>
      <td>0.022208</td>
      <td>0.099196</td>
      <td>0</td>
      <td>0</td>
      <td>9.209647</td>
      <td>35.063898</td>
      <td>NRG1;EGFR</td>
    </tr>
  </tbody>
</table>
</div>

<div class="highlight"><pre><span></span><code><span class="n">gp</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">enr</span><span class="o">.</span><span class="n">results</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Enrichr Analysis of Gene Scores&#39;</span><span class="p">,</span><span class="n">fs</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">top_term</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;../figures/4C.svg&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span> <span class="o">=</span> <span class="s2">&quot;tight&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="png" src="../Figure4_files/Figure4_17_0.png" /></p>
<h2 id="figure-4d-gene-enrichment-analysis-and-visualization">Figure 4D: Gene Enrichment Analysis and Visualization<a class="headerlink" href="#figure-4d-gene-enrichment-analysis-and-visualization" title="Permanent link">&para;</a></h2>
<p>Calculates pathway scores for portal hepatocytes using the genes identified from the enrichment analysis. The results are visualized through PCA projections, highlighting the pathway score and log dose relationships, and saved as figures for further analysis.</p>
<div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;dark&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="n">hep_adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Hepatocytes - portal&quot;</span><span class="p">]</span>
<span class="n">top_gs</span> <span class="o">=</span> <span class="n">enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;Genes&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
<span class="n">pathway</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s2">&quot;Term&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">top_gs</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">top_gs</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes</span><span class="p">(</span><span class="n">hep_adata</span><span class="p">,</span> <span class="n">top_gs</span><span class="p">)</span>
<span class="n">hep_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Log Dose&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">hep_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Dose&quot;</span><span class="p">]]</span>
<span class="c1"># sc.tl.pca(hep_adata)</span>
<span class="c1"># sc.pl.pca(hep_adata, color = &quot;score&quot;, color_map = &quot;flare&quot;, title = &quot;AHR Pathway Score&quot;)</span>
<span class="c1"># sc.pl.pca(hep_adata, color = &quot;Log Dose&quot;, color_map = &quot;flare&quot;)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">hep_adata</span><span class="p">)</span>
</code></pre></div>
<p><div class="highlight"><pre><span></span><code><span class="n">hep_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Log Score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">hep_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
</code></pre></div>
Generates a PCA plot of single-cell expression data for portal hepatocytes, colored by fatty acid oxidation pathway scores.</p>
<div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">hep_adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s2">&quot;flare&quot;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="n">pathway</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;../Figures/Figure3D2.svg&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="png" src="../Figure4_files/Figure4_22_1.png" /></p>
<p>Creates a PCA plot of portal hepatocytes colored by log-transformed TCDD dose values.
<div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">hep_adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;Log Dose&quot;</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s2">&quot;flare&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;../Figures/Figure3D1.svg&quot;</span><span class="p">)</span>
</code></pre></div></p>
<p><img alt="png" src="../Figure4_files/Figure4_23_0.png" /></p>
<h2 id="figure-4e-logistic-fit-of-median-pathway-scores">Figure 4E: Logistic Fit of Median Pathway Scores<a class="headerlink" href="#figure-4e-logistic-fit-of-median-pathway-scores" title="Permanent link">&para;</a></h2>
<p>A logistic curve is fitted to the median pathway scores of fatty acid oxidation across doses, and the Mean Absolute Error (MAE) is calculated to evaluate the fit. The visualization highlights the dose-dependent response in portal hepatocytes.</p>
<div class="highlight"><pre><span></span><code><span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;darkgrid&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">sigmoid</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">x0</span><span class="p">,</span> <span class="n">b</span><span class="p">):</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">L</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">k</span><span class="o">*</span><span class="p">(</span><span class="n">x</span><span class="o">-</span><span class="n">x0</span><span class="p">)))</span><span class="o">+</span><span class="n">b</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">y</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">scipy.optimize</span><span class="w"> </span><span class="kn">import</span> <span class="n">curve_fit</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">pathway</span><span class="p">:</span><span class="n">hep_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">],</span> <span class="s2">&quot;Dose&quot;</span><span class="p">:</span><span class="n">hep_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Dose&quot;</span><span class="p">]})</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Dose&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">y</span> <span class="o">=</span> <span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Dose&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">d</span><span class="p">][</span><span class="n">pathway</span><span class="p">]</span><span class="o">.</span><span class="n">median</span><span class="p">()</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Log Dose&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;Dose&quot;</span><span class="p">])</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;Log Score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">pathway</span><span class="p">])</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;Dose&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">pathway</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">bounds</span> <span class="o">=</span> <span class="p">((</span><span class="mf">1.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">,</span> <span class="mf">0.</span><span class="p">,</span><span class="o">-</span><span class="mf">5.</span><span class="p">),</span> <span class="p">(</span><span class="mf">15.</span><span class="p">,</span><span class="mf">50.</span><span class="p">,</span> <span class="mf">30.</span><span class="p">,</span> <span class="mf">5.</span><span class="p">))</span>
<span class="n">p0</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">]</span>
<span class="n">popt</span><span class="p">,</span> <span class="n">pcov</span> <span class="o">=</span> <span class="n">curve_fit</span><span class="p">(</span><span class="n">sigmoid</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">p0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;trf&#39;</span><span class="p">,</span> <span class="n">maxfev</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span> <span class="n">bounds</span> <span class="o">=</span> <span class="n">bounds</span><span class="p">)</span>
<span class="n">r2</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">sigmoid</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="o">*</span><span class="n">popt</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">sigmoid</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="o">*</span><span class="n">popt</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MAE = </span><span class="si">{</span><span class="n">r2</span><span class="si">:</span><span class="s2">.</span><span class="si">{</span><span class="mi">2</span><span class="si">}}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mf">1.7</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pathway</span><span class="si">}</span><span class="s2"> Score&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;../Figures/Figure3E.svg&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span> <span class="o">=</span> <span class="s2">&quot;tight&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="png" src="../Figure4_files/Figure4_28_0.png" /></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "search.highlight", "search.share", "search.suggest", "content.code.annotate", "content.code.copy"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>